Open Domain Checkout web-view in Site Creation

ovitrif opened the pull request:
Resolves partially #17905

This PR:

- ★ Implements the logic to reuse the existing domain registration checkout and open it in a web-view
- ⊕ Adds API call to create shopping cart for the paid domain product before checkout
- ⊜ Refactors the vm for the progress screen to support returning to it after back navigation from checkout

## To Test

#### Prerequisite

<details><summary>◐ Toggle the <code>SiteCreationDomainPurchasingFeatureConfig</code> flag</summary>

1. Go to `Me` → `App Settings` → `Debug settings`
2. Scroll to the `Features in development` section
3. Tap on the item corresponding to the flag
4. Tap `RESTART THE APP` button
</details>

### ● Treatment Variation

- **Verify** Checkout is opened and filled in with correctly for domain purchases in site creation
  - **Expect** discounts to be applied on Checkout screen
- **Verify** Back navigation from Checkout is supported
  - **Expect** to be able to create a site when resuming with a free domain
  - **Expect** the previously created wpcom blog to be reused when resuming with same domain
  - **Expect** Checkout is opened and filled in correctly when resuming with a different paid domain

### ○ Control Variation

- **Verify** no regression in the site creation flow: expect no issues when creating a new site with a free domain

### Preview

| ● |
| - |
| <video src="https://user-images.githubusercontent.com/4588074/228221635-f6b76ecc-67e3-4827-9203-0b32227c569a.mp4" style="max-width: 315px;"></video> |

> **Note**
> The recording shows a preview of all tests passing for the **● Treatment Variation**.

## Regression Notes
1. Potential unintended areas of impact
  `SITE CREATION` → `PROGRESS`

2. What I did to test those areas of impact (or what existing automated tests I relied on)
  Manual testing and unit testing

3. What automated tests I added (or what prevented me from doing so)
  Unit tests in `SiteCreationMainVMTest` and `SiteCreationProgressViewModelTest`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Update: add supportsPrivacy to domain model

ovitrif committed:
Refactor: remove unneeded suppress lint of ParcelCreator

ovitrif committed:
Refactor: remove unnecessary stubbing on supportsPrivacy

ovitrif committed:
Add: add paid domain product to shopping cart after creating the blog

ovitrif committed:
Test: shopping cart logic

ovitrif committed:
Refactor: remove unused remote site id from state

ovitrif committed:
Refactor: cleanup logic to pre-complete site title task for quickstart

ovitrif committed:
Refactor: add empty lines between injected props

ovitrif committed:
Refactor: remove duplicate test

ovitrif committed:
Refactor: assert generic error uniformly

ovitrif committed:
Add: emit checkout details on cart created successfully

ovitrif committed:
Add: open checkout on cart created successfully

ovitrif committed:
Refactor: rename DomainRegistered to DomainRegistrationPurchased

ovitrif committed:
Add: support back navigation from checkout web-view

ovitrif committed:
Refactor: fix non-uniform naming for checkout event(s)

ovitrif committed:
Test: onCartCreated

ovitrif committed:
Refactor: force Created.InCart contract on checkout result

ovitrif committed:
Test: on checkout result

ovitrif committed:
Refactor: imply site was not fetched in result name

ovitrif committed:
Refactor: replace `when` with `check` to enforce dialog contract

ovitrif committed:
Test: restart progress when returning from checkout

ovitrif committed:
Refactor: fix line length lint issue

ovitrif committed:
Refactor: cleanup test fixtures code

ovitrif committed:
Refactor: remove newline at start of file

ovitrif committed:
Refactor: replace "updates state with result" to "updates result"

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationActivity.kt:
```diff
@@ -122,9 +130,12 @@ class SiteCreationActivity : LocaleAwareActivity(),
+                        intent.putExtra(SitePickerActivity.KEY_SITE_CREATED_BUT_NOT_FETCHED, true)
+                        Triple(true, null, isSiteTitleTaskComplete)
+                    }
+                    is DomainRegistrationPurchased -> {
```
❓ Any particular reason for not using something like the snippet below?
```kotlin
is CreatedButNotFetched, is DomainRegistrationPurchased -> {
    // code here
}
```

ovitrif replied:
Good question @thomashorta, I might have missed to remodel this after I confirmed that everything works via testing.
Resolved in: 61198775

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt:
```diff
@@ -74,34 +81,42 @@ class SiteCreationProgressViewModel @Inject constructor(
+        if (siteCreationState.result is CreatedButNotFetched.InCart) {
+            // reuse the previously blog when returning with the same domain
+            if (siteCreationState.domain == this.siteCreationState.domain) {
+                createCart(siteCreationState.result.remoteId, siteCreationState.result.siteSlug)
```
❓ Just a question/confirmation here: both variables have the same name (`siteCreationState`), is this line using the correct one? (it's using from the parameter)

ovitrif replied:
I'm aware of it and that's why I used `this.` when I wanted the value cached in the VM 👍🏻

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/progress/SiteCreationProgressViewModel.kt:
```diff
@@ -159,6 +180,28 @@ class SiteCreationProgressViewModel @Inject constructor(
+        } else {
+            AppLog.d(LOG_TAG, "Successful cart creation: ${event.cartDetails}")
+            _onCartCreated.postValue(CheckoutDetails(site, domain.domainName))
+            domainsRegistrationTracker.trackDomainsPurchaseWebviewViewed(site, isSiteCreation = true)
```
🔍 Nitpicking: I feel like this event should be tracked closer to the action that actually starts the WebView like [here](https://github.com/wordpress-mobile/WordPress-Android/pull/18177/files#diff-56921dabf3cd181875890b198607814bd8cd3c944dd8e3d46cab28d6e63b738fR300) (`showDomainCheckout` from the other VM), since it seems a bit out of place to track "webViewViewed" in the `onCartCreated` event. Wdyt?

ovitrif replied:
Good point, I reckon I forgot it since the initial copy-pasta "work" 😂.
Resolved in d7a4cd31

thomashorta reviewed (commented):
Overall looks good to me, I just added a few points before I approve.

Also, everything worked while testing but the "sale" domain price shown in the domain picker didn't match the checkout page, see screenshots:
Picker | Checkout
--- | ---
![domain-price-sale-picker](https://user-images.githubusercontent.com/5091503/228583562-d0f2f0c3-11e9-4197-8f9d-a6786fc14b1c.png) | ![domain-price-sale-checkout](https://user-images.githubusercontent.com/5091503/228583602-2489433e-9ac0-4351-bbfd-96eeda488f2a.png)

Also, another feedback unrelated to this PR's work is that the UI is a bit confusing since the selected domain is not highlighted in any way.

ovitrif committed:
Refactor: optimize result model to 3 variants

ovitrif committed:
Refactor: track checkout webview viewed event in main vm

ovitrif reviewed (commented):
@thomashorta Thanks a lot for the review and your remarks 🙇🏻 

I added changes to resolve all concerns you had over the code.

ovitrif commented:
> Also, another feedback unrelated to this PR's work is that the UI is a bit confusing since the selected domain is not highlighted in any way.

👋🏻 @thomashorta 
I totally agree with you on this one, unfortunately we still lack a UI design for highlighting the selected domain, and we are already at the 4th design iteration… 🤦🏻 

Today I collaborated with the developers of the project team and came up with a UI design proposal ourselves.

However, the implementation was already foreseen to be tackled separately, in task:
- #17901

ovitrif commented:
> Also, everything worked while testing but the "sale" domain price shown in the domain picker didn't match the checkout page, see screenshots:

Nice find 🙇🏻 @thomashorta .

Thank you for testing discounts with a different locale. I never encountered this issue with €, but also never tested before with other currencies. I plan to fix this before merging the PR.

thomashorta reviewed (approved):
> Thank you for testing discounts with a different locale. I never encountered this issue with €, but also never tested before with other currencies. I plan to fix this before merging the PR.

As we talked, I think this shouldn't block this PR and has already been added to the tasks in the main issue, so I am approving this one!

Thanks for the contribution and for addressing the review! :shipit:

ovitrif commented:
> > Thank you for testing discounts with a different locale. I never encountered this issue with €, but also never tested before with other currencies. I plan to fix this before merging the PR.
> 
> As we talked, I think this shouldn't block this PR and has already been added to the tasks in the main issue, so I am approving this one!
> 
> Thanks for the contribution and for addressing the review! :shipit:


Posting this for reference & backtracking reasons:
After our DM chat, we concluded it's ok to solve this as part of the main task, where I added a subtask for it:
- https://github.com/wordpress-mobile/WordPress-Android/issues/17905

Therefore I will proceed to merge this PR to the feature branch and make sure the issue you surfaced gets fixed before we merge the whole checkout-related work to `trunk`.

Thank you for reviewing, testing, and helping me with this work @thomashorta  🙇🏻
